File: portfolio/ynab-clarity-ios/YNABClarity/Engine/MetricsEngine.swift
import Foundation

// MARK: - CategoryBalance
// Intermediate value type — bridges YNAB raw data + user's CategoryMapping rules.

struct CategoryBalance {
    let ynabCategoryID: String
    let name: String
    let role: CategoryRole
    /// Monthly target (dollars) — from YNAB goal target, falling back to budgeted/assigned.
    let monthlyTarget: Double
    /// Available balance (dollars) — from YNAB "balance" field.
    let available: Double
    /// Spending activity (dollars, negative = outflow) — from YNAB "activity" field.
    let activityDollars: Double
    /// Day of month the bill is due (0 = not set).
    let dueDay: Int
}

extension CategoryBalance {
    var shortfall: Double { max(0, monthlyTarget - max(0, available)) }

    /// A bill is covered if it still has positive balance (not yet paid), OR if it was
    /// budgeted and a payment has already gone out without going negative (paid this month).
    var isCovered: Bool {
        if available >= monthlyTarget { return true }
        return monthlyTarget > 0 && activityDollars < 0 && available >= 0
    }

    var coverageFraction: Double {
        guard monthlyTarget > 0 else { return 1.0 }
        if isCovered { return 1.0 }
        return min(1.0, max(0, available) / monthlyTarget)
    }
}

// MARK: - MetricsEngine
// All functions are pure (no side effects, no SwiftUI imports).
// Input: arrays of value types. Output: plain Double / Int / Bool.

enum MetricsEngine {

    // MARK: Build from raw YNAB data

    /// Merges YNAB month categories with the user's role mappings.
    /// Filters hidden, deleted, and ignored categories.
    static func buildBalances(
        monthCategories: [YNABMonthCategory],
        mappings: [CategoryMapping]
    ) -> [CategoryBalance] {
        let roleByID = Dictionary(uniqueKeysWithValues: mappings.map { ($0.ynabCategoryID, $0) })
        return monthCategories.compactMap { cat in
            guard !cat.hidden, !cat.deleted else { return nil }
            guard let mapping = roleByID[cat.id] else { return nil }
            let role = mapping.role
            guard role != .ignore else { return nil }
            return CategoryBalance(
                ynabCategoryID: cat.id,
                name: cat.name,
                role: role,
                monthlyTarget: cat.goalTargetDollars ?? cat.budgetedDollars,
                available: cat.balanceDollars,
                activityDollars: cat.activityDollars,
                dueDay: mapping.dueDay
            )
        }
    }

    // MARK: Required obligations

    static func totalRequired(_ balances: [CategoryBalance]) -> Double {
        balances
            .filter { $0.role == .mortgage || $0.role == .bill || $0.role == .essential }
            .reduce(0) { $0 + $1.monthlyTarget }
    }

    static func totalFunded(_ balances: [CategoryBalance]) -> Double {
        balances
            .filter { $0.role == .mortgage || $0.role == .bill || $0.role == .essential }
            .reduce(0) { $0 + min($1.available, $1.monthlyTarget) }
    }

    static func currentShortfall(_ balances: [CategoryBalance]) -> Double {
        max(0, totalRequired(balances) - totalFunded(balances))
    }

    // MARK: Mortgage

    static func mortgageShortfall(_ balances: [CategoryBalance]) -> Double {
        let m = balances.filter { $0.role == .mortgage }
        let target    = m.reduce(0) { $0 + $1.monthlyTarget }
        let available = m.reduce(0) { $0 + $1.available }
        return max(0, target - available)
    }

    static func mortgageCoveredFraction(_ balances: [CategoryBalance]) -> Double {
        let m = balances.filter { $0.role == .mortgage }
        let target = m.reduce(0) { $0 + $1.monthlyTarget }
        guard target > 0 else { return 1.0 }
        let funded = m.reduce(0) { $0 + min($1.available, $1.monthlyTarget) }
        return min(1.0, funded / target)
    }

    // MARK: Bills (non-mortgage required)

    static func billsCoverageFraction(_ balances: [CategoryBalance]) -> Double {
        let b = balances.filter { $0.role == .bill || $0.role == .essential }
        let target = b.reduce(0) { $0 + $1.monthlyTarget }
        guard target > 0 else { return 1.0 }
        let funded = b.reduce(0) { $0 + min($1.available, $1.monthlyTarget) }
        return min(1.0, funded / target)
    }

    // MARK: Safe to spend

    static func safeToSpend(balances: [CategoryBalance]) -> Double {
        let flexAvailable = balances
            .filter { $0.role == .flexible }
            .reduce(0) { $0 + max(0, $1.available) }
        return max(0, flexAvailable - currentShortfall(balances))
    }

    static func safePerDay(balances: [CategoryBalance], daysRemaining: Int) -> Double {
        guard daysRemaining > 0 else { return 0 }
        return safeToSpend(balances: balances) / Double(daysRemaining)
    }

    static func safePerWeek(balances: [CategoryBalance], daysRemaining: Int) -> Double {
        safePerDay(balances: balances, daysRemaining: daysRemaining) * 7
    }

    // MARK: Income

    static func expectedIncomeThisMonth(
        sources: [IncomeSource],
        month: Date,
        calendar: Calendar = .current
    ) -> Double {
        sources.reduce(0.0) { total, source in
            let count = source.occurrencesInMonth(month, calendar: calendar).count
            return total + source.amountDollars * Double(count)
        }
    }

    static func incomeGap(
        balances: [CategoryBalance],
        sources: [IncomeSource],
        month: Date
    ) -> Double {
        max(0, totalRequired(balances) - expectedIncomeThisMonth(sources: sources, month: month))
    }

    // MARK: Salary target

    /// Gross annual salary needed to net `netMonthlyNeeded` per month after tax.
    static func grossAnnualNeeded(netMonthlyNeeded: Double, taxRate: Double) -> Double {
        guard taxRate >= 0, taxRate < 1.0 else { return netMonthlyNeeded * 12 }
        return (netMonthlyNeeded * 12) / (1 - taxRate)
    }

    // MARK: Time helpers

    static func daysRemainingInMonth(from date: Date = Date(), calendar: Calendar = .current) -> Int {
        guard let range = calendar.range(of: .day, in: .month, for: date) else { return 1 }
        let today = calendar.component(.day, from: date)
        return max(1, range.count - today + 1)
    }

    // MARK: Underfunded goals

    static func underfundedGoals(
        monthCategories: [YNABMonthCategory],
        mappings: [CategoryMapping]
    ) -> [GoalStatus] {
        let roleByID = Dictionary(uniqueKeysWithValues: mappings.map { ($0.ynabCategoryID, $0) })
        return monthCategories.compactMap { cat -> GoalStatus? in
            guard !cat.hidden, !cat.deleted else { return nil }
            guard let mapping = roleByID[cat.id], mapping.role != .ignore else { return nil }
            guard let target = cat.goalTargetDollars, target > 0 else { return nil }
            let assigned = cat.budgetedDollars
            let gap = target - assigned
            guard gap > 0.01 else { return nil }
            return GoalStatus(
                categoryName: cat.name,
                goalTarget: target,
                assigned: assigned,
                gap: gap
            )
        }
        .sorted { $0.gap > $1.gap }
    }
}

// MARK: - GoalStatus

struct GoalStatus: Identifiable {
    let id = UUID()
    let categoryName: String
    let goalTarget: Double
    let assigned: Double
    let gap: Double
}

File: portfolio/clarity-hub/src/engines/MetricsEngine.js
// MetricsEngine.js — Pure functions ported from MetricsEngine.swift
// All dollar amounts are in dollars (not cents, not milliunits)

/**
 * Build CategoryBalance array from YNAB month categories + user's role mappings.
 * categoryMappings: [{ ynabCategoryID, ynabCategoryName, ynabGroupName, roleRaw, dueDay }]
 * monthCategories: [{ id, name, budgeted, activity, balance, hidden, deleted, goalTarget, goalType, goalPercentageComplete }]
 *   (amounts in milliunits from YNAB API — divide by 1000 for dollars)
 */
export function buildBalances(monthCategories, categoryMappings) {
  const roleByID = {};
  for (const m of categoryMappings) roleByID[m.ynabCategoryID] = m;

  return monthCategories
    .filter(cat => !cat.hidden && !cat.deleted && roleByID[cat.id] && roleByID[cat.id].roleRaw !== "ignore")
    .map(cat => {
      const mapping = roleByID[cat.id];
      const budgetedDollars = cat.budgeted / 1000;
      const goalTargetDollars = cat.goalTarget != null ? cat.goalTarget / 1000 : null;
      const monthlyTarget = goalTargetDollars != null ? goalTargetDollars : budgetedDollars;
      const available = cat.balance / 1000;
      const activityDollars = cat.activity / 1000;

      const shortfall = Math.max(0, monthlyTarget - Math.max(0, available));
      const isCovered = available >= monthlyTarget || (monthlyTarget > 0 && activityDollars < 0 && available >= 0);
      const coverageFraction = monthlyTarget > 0 ? (isCovered ? 1.0 : Math.min(1.0, Math.max(0, available) / monthlyTarget)) : 1.0;

      return {
        ynabCategoryID: cat.id,
        name: cat.name,
        role: mapping.roleRaw,
        monthlyTarget,
        available,
        activityDollars,
        dueDay: mapping.dueDay || 0,
        shortfall,
        isCovered,
        coverageFraction,
      };
    });
}

export function totalRequired(balances) {
  return balances
    .filter(b => b.role === "mortgage" || b.role === "bill" || b.role === "essential")
    .reduce((sum, b) => sum + b.monthlyTarget, 0);
}

export function totalFunded(balances) {
  return balances
    .filter(b => b.role === "mortgage" || b.role === "bill" || b.role === "essential")
    .reduce((sum, b) => sum + Math.min(b.available, b.monthlyTarget), 0);
}

export function currentShortfall(balances) {
  return Math.max(0, totalRequired(balances) - totalFunded(balances));
}

export function mortgageShortfall(balances) {
  const m = balances.filter(b => b.role === "mortgage");
  const target = m.reduce((s, b) => s + b.monthlyTarget, 0);
  const avail = m.reduce((s, b) => s + b.available, 0);
  return Math.max(0, target - avail);
}

export function mortgageCoveredFraction(balances) {
  const m = balances.filter(b => b.role === "mortgage");
  const target = m.reduce((s, b) => s + b.monthlyTarget, 0);
  if (target <= 0) return 1.0;
  const funded = m.reduce((s, b) => s + Math.min(b.available, b.monthlyTarget), 0);
  return Math.min(1.0, Math.max(0, funded / target));
}

export function billsCoverageFraction(balances) {
  const b = balances.filter(x => x.role === "bill" || x.role === "essential");
  const target = b.reduce((s, x) => s + x.monthlyTarget, 0);
  if (target <= 0) return 1.0;
  const funded = b.reduce((s, x) => s + Math.min(x.available, x.monthlyTarget), 0);
  return Math.min(1.0, Math.max(0, funded / target));
}

export function obligationsCoverageFraction(balances) {
  const slice = balances.filter(b => b.role === "mortgage" || b.role === "bill" || b.role === "essential");
  const target = slice.reduce((s, b) => s + b.monthlyTarget, 0);
  if (target <= 0) return 1.0;
  const funded = slice.reduce((s, b) => s + Math.min(b.available, b.monthlyTarget), 0);
  return Math.min(1.0, Math.max(0, funded / target));
}

export function safeToSpend(balances, toBeBudgeted = 0) {
  const discretionary = balances
    .filter(b => b.role !== "mortgage" && b.role !== "bill" && b.role !== "essential")
    .reduce((sum, b) => sum + Math.max(0, b.available), 0);
  const tbb = Math.max(0, toBeBudgeted);
  return Math.max(0, discretionary + tbb - currentShortfall(balances));
}

export function safePerDay(balances, daysRemaining, toBeBudgeted = 0) {
  if (daysRemaining <= 0) return 0;
  return safeToSpend(balances, toBeBudgeted) / daysRemaining;
}

export function safePerWeek(balances, daysRemaining, toBeBudgeted = 0) {
  return safePerDay(balances, daysRemaining, toBeBudgeted) * 7;
}

// ── Spending from transactions ──

export function outflowSpending(transactions, rangeStart, rangeEndInclusive) {
  let total = 0;
  for (const t of transactions) {
    if (t.deleted || t.transferAccountId) continue;
    const d = new Date(t.date + "T00:00:00");
    if (d >= rangeStart && d <= rangeEndInclusive) {
      const dollars = t.amount / 1000;
      if (dollars < 0) total += -dollars;
    }
  }
  return total;
}

export function spendingYesterday(transactions) {
  const now = new Date();
  const y = new Date(now); y.setDate(y.getDate() - 1);
  y.setHours(0, 0, 0, 0);
  return outflowSpending(transactions, y, y);
}

export function spendingThisWeek(transactions) {
  const now = new Date();
  const today = new Date(now); today.setHours(0, 0, 0, 0);
  // Week starts Monday
  const dayOfWeek = today.getDay();
  const diff = dayOfWeek === 0 ? 6 : dayOfWeek - 1;
  const weekStart = new Date(today); weekStart.setDate(weekStart.getDate() - diff);
  return outflowSpending(transactions, weekStart, today);
}

export function spendingThisMonth(transactions) {
  const now = new Date();
  const today = new Date(now); today.setHours(0, 0, 0, 0);
  const monthStart = new Date(now.getFullYear(), now.getMonth(), 1);
  return outflowSpending(transactions, monthStart, today);
}

// ── Income ──

export function occurrencesInMonth(source, month) {
  const year = month.getFullYear();
  const mo = month.getMonth();
  const daysInMonth = new Date(year, mo + 1, 0).getDate();
  const monthStart = new Date(year, mo, 1);
  const monthEnd = new Date(year, mo, daysInMonth);

  const freq = source.frequencyRaw;
  const amountDollars = source.amountCents / 100;
  const nextPay = new Date(source.nextPayDate + "T00:00:00");
  const payDay = nextPay.getDate();

  if (freq === "monthly") {
    return [new Date(year, mo, Math.min(payDay, daysInMonth))];
  }
  if (freq === "semimonthly") {
    const d1 = Math.min(payDay, daysInMonth);
    const d2 = Math.min(source.secondPayDay || 20, daysInMonth);
    return [new Date(year, mo, d1), new Date(year, mo, d2)].sort((a, b) => a - b);
  }
  if (freq === "biweekly") {
    let anchor = new Date(nextPay);
    while (anchor > monthStart) anchor.setDate(anchor.getDate() - 14);
    const dates = [];
    let current = new Date(anchor);
    while (current <= monthEnd) {
      if (current >= monthStart) dates.push(new Date(current));
      current.setDate(current.getDate() + 14);
    }
    return dates;
  }
  return [];
}

export function expectedIncomeThisMonth(sources, month) {
  return sources.reduce((total, source) => {
    const count = occurrencesInMonth(source, month).length;
    return total + (source.amountCents / 100) * count;
  }, 0);
}

export function incomeGap(balances, sources, month) {
  return Math.max(0, totalRequired(balances) - expectedIncomeThisMonth(sources, month));
}

export function grossAnnualNeeded(netMonthlyNeeded, taxRate) {
  if (taxRate < 0 || taxRate >= 1.0) return netMonthlyNeeded * 12;
  return (netMonthlyNeeded * 12) / (1 - taxRate);
}

export function daysRemainingInMonth(date = new Date()) {
  const daysInMonth = new Date(date.getFullYear(), date.getMonth() + 1, 0).getDate();
  const today = date.getDate();
  return Math.max(1, daysInMonth - today + 1);
}

// ── Underfunded goals ──

export function underfundedGoals(monthCategories, categoryMappings) {
  const roleByID = {};
  for (const m of categoryMappings) roleByID[m.ynabCategoryID] = m;

  return monthCategories
    .filter(cat => {
      if (cat.hidden || cat.deleted) return false;
      const mapping = roleByID[cat.id];
      if (!mapping || mapping.roleRaw === "ignore") return false;
      const target = cat.goalTarget != null ? cat.goalTarget / 1000 : null;
      if (!target || target <= 0) return false;
      const assigned = cat.budgeted / 1000;
      return target - assigned > 0.01;
    })
    .map(cat => {
      const target = cat.goalTarget / 1000;
      const assigned = cat.budgeted / 1000;
      return {
        ynabCategoryID: cat.id,
        categoryName: cat.name,
        goalTarget: target,
        assigned,
        gap: target - assigned,
      };
    })
    .sort((a, b) => b.gap - a.gap);
}
